src/java.base/share/classes/javax/crypto/KDF.java

```diff
@@ -0,0 +1,631 @@
+     * @param kdfParameters
+     *     the {@code KDFParameters} used to configure the derivation
+     *     algorithm or {@code null} if no additional parameters are provided
+     * @param provider
+     *     the provider to use for this key derivation
+     *
+     * @return a {@code KDF} object
+     *
+     * @throws NoSuchAlgorithmException
+     *     if the specified provider does not support the specified KDF
+     *     algorithm
+     * @throws InvalidAlgorithmParameterException
+     *     if the {@code KDFParameters} is an invalid value
+     * @throws NullPointerException
+     *     if the {@code algorithm} or {@code provider} is {@code null}
+     */
+    public static KDF getInstance(String algorithm,
+                                  KDFParameters kdfParameters,
+                                  Provider provider)
+        throws NoSuchAlgorithmException, InvalidAlgorithmParameterException {
+        Objects.requireNonNull(algorithm, "null algorithm name");
+        Objects.requireNonNull(provider, "provider may not be null");
+        try {
+            Instance instance = GetInstance.getInstance("KDF", KDFSpi.class,
+                                                        algorithm,
+                                                        kdfParameters,
+                                                        provider);
+            if (!JceSecurity.canUseProvider(instance.provider)) {
+                String msg = "JCE cannot authenticate the provider "
+                             + instance.provider.getName();
+                throw new SecurityException(msg);
+            }
+            return new KDF((KDFSpi) instance.impl, instance.provider, algorithm,
+                           kdfParameters);
+
+        } catch (NoSuchAlgorithmException nsae) {
+            return handleException(nsae);
+        }
+    }
+
+    private static KDF handleException(NoSuchAlgorithmException e)
+        throws NoSuchAlgorithmException, InvalidAlgorithmParameterException {
+        Throwable cause = e.getCause();
+        if (cause instanceof InvalidAlgorithmParameterException) {
+            throw (InvalidAlgorithmParameterException) cause;
+        }
+        throw e;
+    }
+
+    /**
+     * Derives a key, returned as a {@code SecretKey}.
+     * <p>
+     * The {@code deriveKey} method may be called multiple times on a particular
+     * {@code KDF} instance.
+     *
+     * @param alg
+     *     the algorithm of the resultant {@code SecretKey} object
+     * @param kdfParameterSpec
+     *     the object describing the inputs to the derivation function
+     *
+     * @return a {@code SecretKey} object corresponding to a key built from the
+     *     KDF output and according to the derivation parameters. Rather than
+     *     permitting a {@code null} return value here, implementations will
+     *     handle cases where a value cannot be returned by throwing an
+     *     appropriate {@code Exception}.
```

Comment:
We don't need to specify that `null` is not a valid return value in APIs, it is assumed *unless* you specify that `null` is a valid return value. Otherwise, applications would have to defensively check for `null` for every method where it is not specified, which would be impractical. I think the text above is in the wrong place. If there are situations where an implementation should throw an exception, then it should be specified in the @throws blocks. Is this a case where something has gone seriously wrong and a `ProviderException` should be thrown?

Reply:
Ignore my suggestion. I'm convinced now that unless specified we should consider the return value non nullable. Sorry.

Reply:
resolved.